Publish Microsoft.AzureStackHCI 2021-07-01-preview

Arkanayan opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
 **March 3rd 2022**
3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
**March 3rd 2022**
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

JocelynBerrendonner committed:
Adds base for updating Microsoft.AzureStackHCI from version preview/2021-01-01-preview to version 2021-07-01-preview

JocelynBerrendonner committed:
Updates readme

JocelynBerrendonner committed:
Updates API version in new specs and examples

JocelynBerrendonner committed:
Publishing the 2021-07-01-preview API for self-service VM on HCI

Arkanayan committed:
Fix some linter errors

Arkanayan committed:
Update readme to include 2021-07

Arkanayan committed:
Fix errors and run prettier

Arkanayan committed:
Fix readme.go.md

Arkanayan committed:
Fix SSH description

mentat9 commented:
@Arkanayan - Set this PR to Open and send email to the current ARM API Review oncall when you are ready for ARM review.

mentat9 commented:
@Arkanayan - Please study the materials and attest to that fact by checking the box in the submission form:
**[ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.**

Arkanayan commented:
> @Arkanayan - Please study the materials and attest to that fact by checking the box in the submission form: **[ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.**

Done. @mentat9 

Thanks!

mentat9 commented:
@Arkanayan - I don't see any changes in this PR other than moving from 2021-12-01 to 2022-03-01-preview, so ARM review is not required. I've removed the label.

Note: in the future if you do need ARM review, you will need to first study the materials linked and attest to that by checking this box:
**[ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.**

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+        },
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/SubscriptionIdParameter"
```
>v2/types.json [](http://example.com/codeflow?start=69&length=13)

Consider moving to v3/types.json. Comment applies throughout this PR.

JocelynBerrendonner replied:
Added this workitem to this pending task: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+              "$ref": "#/definitions/galleryimages"
+            }
+          },
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

PATCH doesn't return 201.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+          },
+          "description": "operating system type that the gallery image uses. Expected to be linux or windows"
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

Should be string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+          "type": "string",
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
```
>status [](http://example.com/codeflow?start=9&length=6)

Does this property have a fixed set of values? If so, it should be a string enum.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+      "type": "object",
+      "x-ms-azure-resource": true,
+      "x-ms-client-flatten": true,
+      "properties": {
```
>properties [](http://example.com/codeflow?start=7&length=10)

You can bring in standard definitions for id, name, type, location, tags, and systemData by using the TrackedResource type in common-types v3/types.json

mentat9 replied:
Any reason you chose to use v1/types.json and keep systemData separate? You should be able to use v3/types.json and remove the separate definition for systemData.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,439 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

This belongs in the property bag.

JocelynBerrendonner replied:
+1 this shouldn't be here

JocelynBerrendonner replied:
Agreed. And, this should *not* be a breaking change. @Arkanayan ^

Arkanayan replied:
Resolved.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+              "$ref": "#/definitions/networkinterfaces"
+            }
+          },
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

Remove 201.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+          "$ref": "#/definitions/InterfaceDNSSettings",
+          "description": "DNS Settings for the interface"
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

Should be string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+          "type": "string",
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
```
>status [](http://example.com/codeflow?start=9&length=6)

Sounds like fixed set of values == string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+              "type": "string",
+              "description": "PrivateIPAddress - Private IP address of the IP configuration."
+            },
+            "privateIPAllocationMethod": {
```
>privateIPAllocationMethod [](http://example.com/codeflow?start=13&length=25)

Should be string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+            "subnet": {
+              "type": "object",
+              "properties": {
+                "id": {
```
>id [](http://example.com/codeflow?start=17&length=2)

Recommend the name resourceId for ID references.

JocelynBerrendonner replied:
Aggreed. The API was published to our customers with the "id" property name (this API was copied from RPSaaSdev, where it was reviewed before). In order to avoid introducing any breaking change, we'd like to keep "id" in this version of the API and fix this in a future revision. Does this plan sound good?

I added this workitem to the pending task here: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+      },
+      "description": "InterfaceIPConfiguration iPConfiguration in a network interface."
+    },
+    "networkinterfaces": {
```
>networkinterfaces [](http://example.com/codeflow?start=5&length=17)

Consider using common-types TrackedResource definition.

mentat9 replied:
Same question here: TrackedResource from v3/types.json == TrackedResource from v1/types.json + systemData from v2/types.json.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,480 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

Move to property bag.

JocelynBerrendonner replied:
Agreed. @Arkanayan ^

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/storageContainers.json:
```diff
@@ -0,0 +1,411 @@
+              "$ref": "#/definitions/storagecontainers"
+            }
+          },
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

Remove 201

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/storageContainers.json:
```diff
@@ -0,0 +1,411 @@
+          "type": "string",
+          "description": "location of the image the gallery image should be created from"
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/storageContainers.json:
```diff
@@ -0,0 +1,411 @@
+          "type": "string",
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
```
>status [](http://example.com/codeflow?start=9&length=6)

string enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/storageContainers.json:
```diff
@@ -0,0 +1,411 @@
+      "type": "object",
+      "x-ms-azure-resource": true,
+      "x-ms-client-flatten": true,
+      "properties": {
```
>properties [](http://example.com/codeflow?start=7&length=10)

Recommend TrackedResource definition from common-types

mentat9 replied:
v3/types.json brings in systemData automatically.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/storageContainers.json:
```diff
@@ -0,0 +1,411 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

Move to property bag

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+              "$ref": "#/definitions/virtualharddisks"
+            }
+          },
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

Remove 201

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+          "format": "int64",
+          "description": "diskSizeBytes - size of the disk in GB"
+        },
+        "dynamic": {
```
>dynamic [](http://example.com/codeflow?start=9&length=7)

Prefer string enum vs. boolean. Ref: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review). Could use sizingType: { Static, Dynamic }

JocelynBerrendonner replied:
The API was published to our customers with the "boolean" type (this API was copied from RPSaaSdev, where it was reviewed before). In order to avoid introducing any breaking change, we'd like to keep "boolean" in this version of the API and fix this in a future revision. Does this plan sound good?

JocelynBerrendonner replied:
Added this workitem to this pending task: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+          "type": "integer",
+          "format": "int32"
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

String enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+          "type": "string",
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
```
>status [](http://example.com/codeflow?start=9&length=6)

String enum.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+      "type": "object",
+      "x-ms-azure-resource": true,
+      "x-ms-client-flatten": true,
+      "properties": {
```
>properties [](http://example.com/codeflow?start=7&length=10)

Consider common-types TrackedResource definition

mentat9 replied:
v3/types.json preferred.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,428 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

Move to property bag

Arkanayan replied:
cc @JocelynBerrendonner

JocelynBerrendonner replied:
@Arkanayan , If I understand properly, we should move these properties to a common json file and do a "$ref" here.

JocelynBerrendonner replied:
Doing so should *as per my understanding* not be a breaking change.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+              "$ref": "#/definitions/virtualmachines"
+            }
+          },
+          "201": {
```
>201 [](http://example.com/codeflow?start=11&length=3)

Remove 201

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+        "hardwareProfile": {
+          "type": "object",
+          "properties": {
+            "vmSize": {
```
>vmSize [](http://example.com/codeflow?start=13&length=6)

Consider string enum.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+        "osProfile": {
+          "type": "object",
+          "properties": {
+            "adminPassword": {
```
>adminPassword [](http://example.com/codeflow?start=13&length=13)

Secrets are not allowed in GET payloads. If it needs to be retrievable, it must be via a separate POST action.

JocelynBerrendonner replied:
Good point, we should remove this from the GET operation.

JocelynBerrendonner replied:
@Arkanayan ^

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+            "linuxConfiguration": {
+              "type": "object",
+              "properties": {
+                "disablePasswordAuthentication": {
```
>disablePasswordAuthentication [](http://example.com/codeflow?start=17&length=29)

Consider string enum, e.g. authenticationPolicy { AllowPassword, DisallowPassword }, or similar.

JocelynBerrendonner replied:
As mentioned in a previous comment, we cannot change this without breaking customers for this API (the API was published in RPSaaSDev a few months ago and we forgot to port that API to main). That said, in the long run, I agree with changing this to string enum. We will pick this up in the next version of the API.

JocelynBerrendonner replied:
Added this workitem to this pending task: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+                      "items": {
+                        "type": "object",
+                        "properties": {
+                          "keyData": {
```
>keyData [](http://example.com/codeflow?start=27&length=7)

x-ms-secret?

mentat9 replied:
Presuming this also doesn't come back on GET. Would be good to include the x-ms-mutability extension to document that fact.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+              },
+              "description": "LinuxConfiguration - linux specific configuration values for the virtual machine"
+            },
+            "osType": {
```
>osType [](http://example.com/codeflow?start=13&length=6)

String enum?

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+            "windowsConfiguration": {
+              "type": "object",
+              "properties": {
+                "enableAutomaticUpdates": {
```
>enableAutomaticUpdates [](http://example.com/codeflow?start=17&length=22)

Suggest string enum, e.g. updatePolicy: { Automatic, Manual }

JocelynBerrendonner replied:
As mentioned in a previous comment, we cannot change this without breaking customers for this API (the API was published in RPSaaSDev a few months ago and we forgot to port that API to main). That said, in the long run, I agree with changing this to string enum. We will pick this up in the next version of the API.

JocelynBerrendonner replied:
Added this workitem to this pending task: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+        "securityProfile": {
+          "type": "object",
+          "properties": {
+            "enableTPM": {
```
>enableTPM [](http://example.com/codeflow?start=13&length=9)

String enum: TPM: { Enabled, Disabled }.

JocelynBerrendonner replied:
Added this workitem to this pending task: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+          },
+          "description": "StorageProfile - contains information about the disks and storage information for the virtual machine"
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

String enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+      "type": "object",
+      "x-ms-client-flatten": true,
+      "x-ms-azure-resource": true,
+      "properties": {
```
>properties [](http://example.com/codeflow?start=7&length=10)

Consider TrackedResources from common-types.

mentat9 replied:
v3/types.json preferred.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

Move to property bag.

JocelynBerrendonner replied:
Agreed. @Arkanayan ^

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,801 @@
+          "items": {
+            "type": "object",
+            "properties": {
+              "id": {
```
>id [](http://example.com/codeflow?start=15&length=2)

Suggest resourceId for the name of this property.

JocelynBerrendonner replied:
Agreed. That said, the API was published to our customers with the "id" property name (this API was copied from RPSaaSdev, where it was reviewed before). In order to avoid introducing any breaking change, we'd like to keep "id" in this version of the API and fix this in a future revision. Does this plan sound good?

I added this workitem to the pending task here: https://dev.azure.com/msazure/msk8s/_boards/board/t/ARC-HCI/Stories/?workitem=12679581

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+                "type": "object",
+                "x-ms-client-flatten": true,
+                "properties": {
+                  "addressPrefix": {
```
String enum

JocelynBerrendonner replied:
As mentioned in a previous comment, we cannot change this without breaking customers for this API (the API was published in RPSaaSDev a few months ago and we forgot to port that API to main). That said, in the long run, I agree with changing this to string enum. We will pick this up in the next version of the API.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+                    },
+                    "description": "AddressPrefixes - List of address prefixes for the subnet."
+                  },
+                  "ipAllocationMethod": {
```
String enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+                    "items": {
+                      "type": "object",
+                      "properties": {
+                        "ID": {
```
>ID [](http://example.com/codeflow?start=25&length=2)

Suggest resourceId. At minimum it should be camelCase (== id)

Arkanayan replied:
We expect name of the resource here. Updated the description.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+                  "routeTable": {
+                    "type": "object",
+                    "properties": {
+                      "id": {
```
>id [](http://example.com/codeflow?start=23&length=2)

eTag?

JocelynBerrendonner replied:
So, in order to avoid breaking existing changes, we will want to keep it as "ID" for now. We can change it in the next version of the API.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+          "description": "Subnet - list of subnets under the virtual network",
+          "x-ms-identifiers": []
+        },
+        "provisioningState": {
```
>provisioningState [](http://example.com/codeflow?start=9&length=17)

String enum

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+        },
+        "end": {
+          "type": "string",
+          "description": "end of the ip address pool"
```
>end of the ip address pool [](http://example.com/codeflow?start=26&length=26)

First IP/last IP of the pool is slightly more specific language than Start/End of pool.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+        }
+      }
+    },
+    "virtualnetworks": {
```
Consider using TrackedResource definition from common-types.

mentat9 replied:
v3/types.json preferred.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,575 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+        },
+        "extendedLocation": {
```
>extendedLocation [](http://example.com/codeflow?start=9&length=16)

Move to property bag.

Arkanayan replied:
cc @JocelynBerrendonner

JocelynBerrendonner replied:
We should talk to the custom location folks. They should add this to the property bag for every consumer of this code.

JocelynBerrendonner replied:
Actually, let's first add it to a property bag we create. Other teams seem to have done that.

mentat9 commented:
@Arkanayan - I've reviewed this PR for ARM and left some comments.

Arkanayan committed:
Set provisioningState as enum

Arkanayan committed:
Use common-types TrackedResource

Arkanayan committed:
Change vmSize type to enum

Arkanayan committed:
Set adminPassword mutability to [create]

Arkanayan committed:
Set osType as Enum

Arkanayan committed:
Set networkType as enum

Arkanayan committed:
Some changes based on feedback

Arkanayan committed:
Remove storagecontainers

Arkanayan committed:
Add MocVirtualMachineStatus object

Arkanayan committed:
Fix provisioningState enum

Arkanayan committed:
Fix networkinterface id description

Arkanayan committed:
Fix provisioningState in all the resources

Arkanayan committed:
Update description of ipConfigurationReferences.ID

Arkanayan committed:
Add typed status object

Arkanayan committed:
Use common-types v3/types.json objects

Arkanayan committed:
Remove 201 response from update apis

Arkanayan committed:
Add enum PrivateIPAllocationMethod

Arkanayan commented:
@mentat9 I have addressed most of the review comments. As communicated via email, we missed the review process and already published this version. Our customers have been using these non-reviewed API for a few weeks now. The breaking changes will negatively impact them. We'll maintain the current preview for a minimum of 90 days and fix the rest of the suggestions in the next version.

mentat9 commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,854 @@
+          }
+        },
+        "status": {
+          "$ref": "#/definitions/MOCVirtualMachineStatus"
```
>#/definitions/MOCVirtualMachineStatus" [](http://example.com/codeflow?start=19&length=38)

Nice improvement.

mentat9 commented:
> @mentat9 We have addressed most of the review comments. As communicated via email, we missed the review process and already published this version. Our customers have been using these non-reviewed API for a few weeks now. The breaking changes will negatively impact them. We'll maintain the current preview for a minimum of 90 days and fix the rest of the suggestions in the next version.

Got it, thanks @Arkanayan. I've reviewed your updates, resolved the related comments, and added/updated a few of the existing comments. Since you have already published your APIs you don't need ARM signoff for this PR, as it's just documenting what's already published, so I've removed the WaitForARMFeedback label, which will allow this PR to merge. You will get ARM signoff on your next API version (apply the feedback from this PR on your next one).

Arkanayan commented:
Bump

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+          },
+          "description": "operating system type that the gallery image uses. Expected to be linux or windows"
+        },
+        "provisioningState": {
```
Can you make this readonly as well, please?

JocelynBerrendonner replied:
(same comment applies to all the other resources: networkinterface, virtualharddisk, virtualmachines, ...)

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
+          "$ref": "#/definitions/MOCGalleryImageStatus"
```
I am curious, why is it "MOCGalleryImageStatus" as opposed to "GalleryImageStatus" ?

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+    },
+    "MOCGalleryImageStatus": {
+      "type": "object",
+      "description": "MOCGalleryImageStatus defines the desired state of MOCGalleryImage",
```
Same here. I am curious about why we expose "MOC" in the ARM API. MOC is an internal concept that I would expect to remain hidden from the swagger.

JocelynBerrendonner replied:
Can you also fix the comment? it is not about the desired state here, it is about the actual state of the gallery image

pathuang-sia replied:
could switch out "desired" to "observed"

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+          "type": "string",
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
```
Can you make this one readonly as well, please?

(same comment applies to all the other resources: networkinterface, virtualharddisk, virtualmachines, ...)

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,414 @@
+          "description": "name of the object to be used in moc"
+        },
+        "status": {
+          "$ref": "#/definitions/MOCVirtualHardDiskStatus"
```
Same comment applies here. We should drop all the references to "MOC" since this isn't a customer facing concept.

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,584 @@
+        },
+        "start": {
+          "type": "string",
+          "description": "First IP of the pool"
```
[Nitpick] First IP in the pool

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,584 @@
+        },
+        "end": {
+          "type": "string",
+          "description": "Last IP of the pool"
```
[Nitpick] Last IP in the pool

pathuang-sia commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,584 @@
+    "MOCVirtualNetworkStatus": {
+      "type": "object",
+      "description": "MOCVirtualNetworkStatus defines the desired state of MOCVirtualNetwork",
+      "x-mx-client-flatten": true,
```
typo: x-ms-client-flatten

pathuang-sia commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+    "MOCGalleryImageStatus": {
+      "type": "object",
+      "description": "MOCGalleryImageStatus defines the desired state of MOCGalleryImage",
+      "x-mx-client-flatten": true,
```
typo: x-ms-client-flatten

pathuang-sia commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,474 @@
+    "MOCNetworkInterfaceStatus": {
+      "type": "object",
+      "description": "MOCNetworkInterfaceStatus defines the desired state of MOCNetworkInterface",
+      "x-mx-client-flatten": true,
```
typo: x-ms-client-flatten

pathuang-sia commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,414 @@
+    "MOCVirtualHardDiskStatus": {
+      "type": "object",
+      "description": "MOCVirtualHardDiskStatus defines the desired state of MocVirtualHardDisk",
+      "x-mx-client-flatten": true,
```
typo

pathuang-sia commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+          },
+          "description": "operating system type that the gallery image uses. Expected to be linux or windows"
+        },
+        "provisioningState": {
```
You might be mixing up provisioningState with provisioningStatus. The former is set by RPaaS, which can have values such as "Accepted", "Created"... etc., while the latter is determined by status of the corresponding moc custom resource.

Arkanayan replied:
I was indeed mixing up these two. I did a search across the repo, found many different variants. I am updating it to contain 
`["Succeeded",
            "Failed",
            "InProgress",
            "Accepted",
            "Deleting"]` . Let me know if this the correct set.

Arkanayan replied:
Should I remove `InProgress`?

JocelynBerrendonner replied:
Let's keep InProgress

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/galleryImages.json:
```diff
@@ -0,0 +1,425 @@
+          "enum": [
+            "Succeeded",
+            "Failed",
+            "InProgress"
```
RPaaS also returns "Accepted" for this value. We should talk to John Breen and understand what other values RPaaS can return here.

Arkanayan committed:
Add additional provisioningState enum variants

Arkanayan committed:
Fix typo

Arkanayan committed:
Remove MOC reference from object definitions

Arkanayan committed:
Set provisioningState as read only

JocelynBerrendonner reviewed (approved):
LGTM

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/networkInterfaces.json:
```diff
@@ -0,0 +1,477 @@
+      "description": "NetworkInterfaceSpec defines the desired state of NetworkInterface"
+    },
+    "NetworkInterfaceStatus": {
+      "type": "object",
```
This needs to be "readonly"

Arkanayan replied:
Resolved. It seems I missed adding readOnly to outer level objects.

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,587 @@
+      },
+      "description": "VirtualNetworkSpec defines the desired state of VirtualNetwork"
+    },
+    "VirtualNetworkStatus": {
```
This needs to be "readonly"

Arkanayan replied:
Resolved.

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualHardDisks.json:
```diff
@@ -0,0 +1,417 @@
+      "description": "VirtualHardDiskSpec defines the desired state of VirtualHardDisk"
+    },
+    "VirtualHardDiskStatus": {
+      "type": "object",
```
This needs to be "readonly"

Arkanayan replied:
Resolved

Arkanayan committed:
Mark Statuses as readOnly

Arkanayan commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualMachines.json:
```diff
@@ -0,0 +1,848 @@
+      "description": "VirtualMachineSpec defines the desired state of VirtualMachine"
+    },
+    "VirtualMachineStatus": {
+      "type": "object",
```
Resolved

JocelynBerrendonner commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2021-07-01-preview/virtualNetworks.json:
```diff
@@ -0,0 +1,588 @@
+                    "properties": {
+                      "id": {
+                        "type": "string",
+                        "description": "Etag - Gets a unique read-only string that changes whenever the resource is updated."
```
I am not sure the description is correct here. Thoughts @Arkanayan ?

Arkanayan replied:
Yeah, the description is incorrect. I have updated it to "Name of the Route". How does it sound? I have never used this field.

JocelynBerrendonner replied:
let's call it "route ID - ID of the IP route"

Arkanayan replied:
Resolved.

Arkanayan committed:
Change description of routeTable ID

Arkanayan committed:
Put extended location definition in a common file

Arkanayan committed:
Change description of route.ID

Arkanayan committed:
Fix customLocation enum types

Arkanayan committed:
Remove storageContainers examples

StorageContainer is not published in this version

Arkanayan committed:
Fix swagger lintdiff

Arkanayan committed:
Remove 201 response from update examples

fixes RESPONSE_STATUS_CODE_NOT_IN_SPEC [https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/Semantic-and-Model-Violations-Reference.md\#RESPONSE_STATUS_CODE_NOT_IN_SPEC]

Arkanayan committed:
Fix PrettierCheck on common.json

Arkanayan commented:
> > @mentat9 We have addressed most of the review comments. As communicated via email, we missed the review process and already published this version. Our customers have been using these non-reviewed API for a few weeks now. The breaking changes will negatively impact them. We'll maintain the current preview for a minimum of 90 days and fix the rest of the suggestions in the next version.
> 
> Got it, thanks @Arkanayan. I've reviewed your updates, resolved the related comments, and added/updated a few of the existing comments. Since you have already published your APIs you don't need ARM signoff for this PR, as it's just documenting what's already published, so I've removed the WaitForARMFeedback and ARMChangesRequested labels, which will allow this PR to merge. You will get ARM signoff on your next API version (apply the feedback from this PR on your next one).

Hi @mentat9, How do we get this merged? As communicated earlier, we have already published our APIs, this PR just documenting that. There are some pipeline failures (But fixing those will end up in breaking changes). Is this the reason we are not able to merge it?

mentat9 commented:
> > > @mentat9 We have addressed most of the review comments. As communicated via email, we missed the review process and already published this version. Our customers have been using these non-reviewed API for a few weeks now. The breaking changes will negatively impact them. We'll maintain the current preview for a minimum of 90 days and fix the rest of the suggestions in the next version.
> > 
> > 
> > Got it, thanks @Arkanayan. I've reviewed your updates, resolved the related comments, and added/updated a few of the existing comments. Since you have already published your APIs you don't need ARM signoff for this PR, as it's just documenting what's already published, so I've removed the WaitForARMFeedback and ARMChangesRequested labels, which will allow this PR to merge. You will get ARM signoff on your next API version (apply the feedback from this PR on your next one).
> 
> Hi @mentat9, How do we get this merged? As communicated earlier, we have already published our APIs, this PR just documenting that. There are some pipeline failures (But fixing those will end up in breaking changes). Is this the reason we are not able to merge it?
> 
> Thanks.

PR review, approval and merge are owned by the Azure SDK team: the PR assignee is always from that team (@jianyexi in this case).

Arkanayan commented:
Hi @jianyexi, Can we get a signoff?

jianyexi commented:
@Arkanayan could u fix the linter errors: https://github.com/Azure/azure-rest-api-specs/pull/18152/checks?check_run_id=8070780141  and the model validation error: https://github.com/Azure/azure-rest-api-specs/pull/18152/checks?check_run_id=8070780448

Arkanayan committed:
Set long running operation options final state to location

Arkanayan committed:
Add operations examples

Arkanayan commented:
Hi @jianyexi, can you take a look, please? 

The lint and model validation has passed.

Arkanayan commented:
Hi @jianyexi,

The tests have passed.
Can I get a sign off, please?

cc: @JocelynBerrendonner

Arkanayan commented:
bump

Arkanayan committed:
Merge remote-tracking branch 'upstream/main' into user/jocelynb/dev-azurestackhci-Microsoft.AzureStackHCI-2021-07-01-preview

Arkanayan committed:
Fix model validation errors

jianyexi commented on custom-words.txt:
```diff
@@ -2799,7 +2799,15 @@ logalertv1metricmeasurement
 routesets
 wayfind
 wayfinding
+>>>>>>> upstream/main
```
please resolve the git conflict

Arkanayan replied:
Fixed.

Arkanayan committed:
Fix git merge
